### PR TITLE
[Refactor:System] Added API route to addNewTerm

### PR DESCRIPTION
**Pull request by elsume**
### Why is this Change Important & Necessary?
Term creation did not have an API route. This PR allows term creation to utilize the API.

### What is the New Behavior?
The changes will let term creation be used as an API endpoint and will update error and success responses to return JSON so API callers get better responses. Additionally, there is now new cypress testing for the new API endpoint.

### What steps should a reviewer take to reproduce or test the bug or new feature?
Testers can manually make HTTP requests with curl commands to test new API endpoint

### Automated Testing & Documentation
Cypress testing for term creation exists. If CI test passes, then that means there are no errors.

### CODE REVIEW

File: site/app/controllers/HomePageController.php [hunk 1]
```diff
@@ -484,43 +484,71 @@ public function getGroupUsers($group_name = null): MultiResponse {
...
+                $term_id,
+                $term_name,
+                new \DateTime($start_date),
+                new \DateTime($end_date),
+            );
+            $em->persist($term);
+            $em->flush();
+            $this->core->addSuccessMessage("Term added successfully.");
+            return new MultiResponse(
+                JsonResponse::getSuccessResponse(null),
```
Comment (IDzyre) [inline site/app/controllers/HomePageController.php]: I wonder if you should return either the term_id, or the term object itsself here. I think it would be beneficial for testing to see that the object was successfully created.

Comment (elsume) [reply site/app/controllers/HomePageController.php]: I agree, that would be better. I have updated it to return the term object and have also updated the cypress test to expect and verify the term object data.

File: site/cypress/e2e/Cypress-Gradeable/gradeable_upload.spec.js
Comment (IDzyre) [inline site/cypress/e2e/Cypress-Gradeable/gradeable_upload.spec.js:1]: Now that the CI tests are passing, you should create a new file in Cypress-Admin, probably titled something like instructor_api.spec.js, and copy the new tests to it. Once this is merged, you could use the same file to test course creation VIA the API also.

Comment (elsume) [reply site/cypress/e2e/Cypress-Gradeable/gradeable_upload.spec.js:1]: Since term creation is something that requires superuser permission while course creation needs instructor permission, I don't think naming it instructor_api.spec.js would fit. Would something like creation_api.spec.js work instead? This would allow me to add the course creation tests to the file in the future, keeping the POST/create endpoints together. If their tests being together isn't necessary, a clearer name would be term_api.spec.js while I'd name the later file course_api.spec.js. Let me know which one you think works best.

Comment (IDzyre) [reply site/cypress/e2e/Cypress-Gradeable/gradeable_upload.spec.js:1]: You could name it `superuser_api.spec.js` if you think that would be better. I think `creation_api` is a bit vague, and since it is in `Cypress-Admin`, I think superuser in the name would be better.

Comment (NicholasCiuica) [reply site/cypress/e2e/Cypress-Gradeable/gradeable_upload.spec.js:1]: I am wondering what is the benefit of combining the tests for term creation and course creation? I think it would be clearer if they were separate.

**Review by NicholasCiuica — COMMENTED** [review]
I've done a code review for this PR and tested it locally. The cypress test looks good, and besides moving them to their own file, I think this PR is ready for merge.

File: site/cypress/e2e/Cypress-Gradeable/gradeable_upload.spec.js
Comment (elsume) [reply site/cypress/e2e/Cypress-Gradeable/gradeable_upload.spec.js:1]: I think keeping them together would only have benefit if they're the only tests for these kind of API endpoints. Then again, I agree that it would be cleaner and more organized if they were separate. It'll also makes it easier to add future tests related to either endpoint. For now I'll go ahead and create superuser_api.spec.js. When the course creation tests are added, we can see about putting them in a separate instructor_api.spec.js.

**Comment by IDzyre** [comment]
There are arguments for both being separate, and combining them. An argument for combining them, is it is a place to test all Superuser API endpoints in one location, so you can tell that the superuser only endpoints work. If you think of each endpoint as a separate object, then yes, separating them would be better.

I think if you had CI tests for creating courses through the UI, and then also ones for creating courses in the API, that can be separate from creating Terms through the UI, and the API.

**Review by jndlansh — COMMENTED** [review]
The PR looks good to me overall. The error handling and JSON responses are clean, and the test coverage for the edge cases is solid. Approving from my end!

**Comment by NicholasCiuica** [comment]
I ran the spec locally again after it has been moved to superuser_api and have run into an issue on the first test now. Can anyone else reproduce this issue locally? I am not sure why this is happening now, since the test itself wasn't changed during the move. Maybe something to do with beforeEach/afterEach being defined in these files? Or some command run in ci.yml that allows this test to pass only on the remote CI? If this test is supposed to fail locally, that should be documented too.

Screenshot 2026-06-30 111822

**Comment by IDzyre** [comment]
> I ran the spec locally again after it has been moved to superuser_api and have run into an issue on the first test now. Can anyone else reproduce this issue locally? I am not sure why this is happening now, since the test itself wasn't changed during the move. Maybe something to do with beforeEach/afterEach being defined in these files? Or some command run in ci.yml that allows this test to pass only on the remote CI? If this test is supposed to fail locally, that should be documented too.
> 
> Screenshot 2026-06-30 111822

In the current design, this test will only work on the first run, as the term will already have been created. You would need to destroy and up to test it again, so that might be your issue.

File: site/cypress/e2e/Cypress-Admin/superuser_api.spec.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    it('Should create a new term', () => {
+        getApiKey('superuser', 'superuser').then((key) => {
+            cy.request({
+                method: 'POST',
+                url: `${Cypress.config('baseUrl')}/api/terms`,
+                body: {
+                    term_id: 's20',
+                    term_name: 'Spring 2020',
+                    start_date: '2020-01-01',
+                    end_date: '2020-05-31',
```
Comment (IDzyre) [inline site/cypress/e2e/Cypress-Admin/superuser_api.spec.js]: The term_name and id could(should) be "randomly" generated, or something similar to allow for multiple re-runs of this test. Currently, this test only works once, then you have to destroy and up to test it again.

File: site/cypress/e2e/Cypress-Admin/superuser_api.spec.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    it('Should show error when term already exists', () => {
+        getApiKey('superuser', 'superuser').then((key) => {
+            cy.request({
+                method: 'POST',
+                url: `${Cypress.config('baseUrl')}/api/terms`,
+                body: {
+                    term_id: 's26',
+                    term_name: 'Spring 2026',
+                    start_date: '2026-01-02',
+                    end_date: '2026-06-30',
```
Comment (IDzyre) [inline site/cypress/e2e/Cypress-Admin/superuser_api.spec.js]: Instead of using this one, you should use the one you created previously. This course will not exist in the fall, since the term created by the dev system is the current term (f26, s26, etc)

**Review by IDzyre — CHANGES_REQUESTED** [review]
Couple of testing requests, but the API part looks good.

File: site/cypress/e2e/Cypress-Admin/superuser_api.spec.js (hunk 2 above)
Comment (elsume) [reply site/cypress/e2e/Cypress-Admin/superuser_api.spec.js]: I've made the change to using Date.now() for unique ID generation. This will make it so the tests won't conflict with preexisting termIDs.

File: site/cypress/e2e/Cypress-Admin/superuser_api.spec.js (hunk 3 above)
Comment (elsume) [reply site/cypress/e2e/Cypress-Admin/superuser_api.spec.js]: I have changed this test to use the same term data values as the first test so that it will consistently show the error of the term already existing. One note is that this test will now fail if the first test for successful term creation does not pass.

File: site/cypress/e2e/Cypress-Admin/superuser_api.spec.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        });
+    });
+    // if the first test fails, this test will naturally also fail
+    it('Should show error when term already exists', () => {
+        getApiKey('superuser', 'superuser').then((key) => {
+            cy.request({
+                method: 'POST',
+                url: `${Cypress.config('baseUrl')}/api/terms`,
+                body: {
+                    term_id: `test${timestamp}`,
```
Comment (IDzyre) [inline site/cypress/e2e/Cypress-Admin/superuser_api.spec.js]: Now that these are dependent, they should be under the same "it" block

**Review by IDzyre — APPROVED** [review]
Changes look good, can still create term in the UI, and the test looks good. I am going to wait a day to merge this to make sure the failing tests aren't from this PR.

**Review by NicholasCiuica — APPROVED** [review]
I reran the superuser_api cypress test locally again and now it works. Also the current CI failures are either flaky or due to the term change, so they should not be an issue. Nice work!

**Review by dandrecollins07-ctrl — APPROVED** [review]
UI and API both passing tests. All Cypress tests passing. Approving.

**Review by bmcutler — COMMENTED** [review]
I did a brief (not exhaustive) test of the API and it worked for me.
